rosette-scala/src/main/scala/coop/rchain/rosette/Tuple.scala

```diff
@@ -1,14 +1,145 @@
+  *  between two actors in a communication, to introduce sub-structure
+  *  in messages, and as a primitive data structure in the definition
+  *  of other actors
+  */
 case class Tuple(elem: Seq[Ob],
-                 override val entry: Seq[Ob],
+                 override val parent: Ob,
                  override val meta: Ob,
                  override val slot: Seq[Ob])
-    extends Ob
+    extends Ob {
+
+  def makeSlice(offset: Int, n: Int): Tuple =
+    this match {
+      case Tuple.NIL => Tuple.NIL
+      case _ => Tuple(n, this, offset, n)
+    }
+
+  def makeTail(entriesToSkip: Int): Tuple = {
+    val size = this.elem.size - entriesToSkip
+    Tuple(size, this, entriesToSkip, size)
+  }
+
+  def nth(n: Int): Option[Ob] = this.elem.lift(n)
+
+  def setNth(n: Int, ob: Ob): Option[Tuple] =
+    try {
+      Some(Tuple(this.elem.updated(n, ob), this.parent, this.meta, this.slot))
+    } catch {
+      case _: IndexOutOfBoundsException => None
+    }
+
+  def subObject(i: Int, n: Int): Tuple = makeSlice(i, n)
+
+  def accepts(msg: Ctxt): Boolean =
+    if (this == Tuple.NIL) {
+      true
+    } else {
+      this.elem.exists(_.matches(msg))
+    }
+
+  override def matches(msg: Ctxt): Boolean = {
+    val n = this.elem.size
+
+    if (n > 0 && n <= msg.nargs) {
+      if (this.elem.head != msg.trgt) {
+        false
+      } else {
+
+        /*
+         * Skip msg.head, since it is actually the receiver of the
+         * message, i.e.,
+         *
+         * 	(oprn rcvr arg1 arg2 ...)
+         *
+         * results in an argvec of
+         *
+         * 	[rcvr arg1 arg2 ...]
+         */
+
+        this.elem
+          .drop(1)
+          .zip(Stream from 0)
+          .forall {
+            case (e, i) =>
+              msg.arg(i) match {
+                case Some(ob) => ob == e
+                case None => false
+              }
+          }
+      }
+    } else {
+      false
+    }
+  }
+
+  def matches(msg: Tuple): Boolean =
+    if (this == Tuple.NIL) {
+      true
+    } else {
+      val n = this.elem.size
+
+      if (n > 0 && n <= msg.elem.size) {
+
+        this.elem
+          .zip(Stream from 0)
+          .forall {
+            case (e, i) =>
+              if (e != msg.elem(i) && e != Ob.NIV) {
+                // TODO:
+                if (e.isA[Tuple] && msg.elem(i).isA[Tuple]) {
+                  if (e.asInstanceOf[Tuple]
+                        .matches(msg.elem(i).asInstanceOf[Tuple])) {
+                    true
+                  } else false
+                } else false
+              } else true
+          }
+
+      } else false
+    }
+
+}
 
 object Tuple {
+
+  object NIL extends Tuple(null, null, null, null)
+
+  val PLACEHOLDER = new Tuple(Seq(), null, null, null)
+
+  def apply(init: Ob) = new Tuple(Seq(init), null, null, null)
+
+  def apply(t1: Tuple, t2: Tuple): Tuple =
+    new Tuple(t1.elem ++ t2.elem, null, null, null)
+
+  def apply(size: Int,
```

Comment:
This pun on the constructor threw me off when I was reading the code above. How about calling it `slice`?

Reply:
This is more or less a one-to-one translation from the C++ implementation.
While I agree with you that this should be refactored at some point, I think it is easier to refactor it later once we saw all usages of this constructor.